### DISCUSSION TITLE
How to control the number of threads of CmaEsSampler?

### DISCUSSION THREAD

**Discussion by n-tateiwa** [optuna/optuna#4009] [Q&A] [answered]
When I select CmaEsSampler as a sampler of study, usage of CPU reaches 400%, which uses full of my quad-core PC.
What should I do to suppress the number of threads used in CmaEsSampler?

environment of my PC

- optuna: v2.10.1
- python: v3.10.5
- macOS Monterey: v12.6

I used the following code to check above fact.

```python
import optuna

def objective(trial):
    obj = 0.0
    for i in range(1000):
        x = trial.suggest_uniform(f"x{i}", 0, 100)
        obj += x * x
    return obj

sampler = optuna.samplers.CmaEsSampler()
study = optuna.study.create_study(sampler=sampler)
study.optimize(objective, n_trials=1000, n_jobs=1)
```

**Comment by amylase** [comment, marked as answer]
It is because `cmaes` package uses NumPy's dot function, which is backed by a library that exploits multi-thread capability.
You can use `threadpoolctl` to limit thread usage by libraries.

```
import optuna
import threadpoolctl


def objective(trial):
    obj = 0.0
    for i in range(1000):
        x = trial.suggest_uniform(f"x{i}", 0, 100)
        obj += x * x
    return obj


sampler = optuna.samplers.CmaEsSampler()
study = optuna.study.create_study(sampler=sampler)

with threadpoolctl.threadpool_limits(limits=1):
    study.optimize(objective, n_trials=1000, n_jobs=1)
```

**Comment by n-tateiwa** [comment]
Thank @amylase  for your comment !
I confirm to above code suppresses the number of threads (and usage of CPU).
